WhateverGreen/kern_igfx_backlight.cpp

```diff
@@ -0,0 +1,452 @@
+	// We only need to patch the framebuffer driver
+	requiresPatchingFramebuffer = true;
+	
+	// We need R/W access to MMIO registers
+	requiresMMIORegistersReadAccess = true;
+	requiresMMIORegistersWriteAccess = true;
+}
+
+void IGFX::BacklightSmoother::deinit() {
+	// `BacklightSmoother::processKernel()` guarantees that all pointers are nullptr on failure
+	if (workloop != nullptr && eventSource != nullptr) {
+		workloop->removeEventSource(eventSource);
+		OSSafeReleaseNULL(eventSource);
+		OSSafeReleaseNULL(workloop);
+	}
+	BrightnessRequestQueue::safeDeleter(queue);
+	OSSafeReleaseNULL(owner);
+}
+
+void IGFX::BacklightSmoother::processKernel(KernelPatcher &patcher, DeviceInfo *info) {
+	// Check the boot argument and the device property
+	enabled = checkKernelArgument("-igfxbls");
+	if (!enabled)
+		enabled = info->videoBuiltin->getProperty("enable-backlight-smoother") != nullptr;
+	if (!enabled)
+		return;
+	
+	// Fetch user configurations
+	if (WIOKit::getOSDataValue(info->videoBuiltin, "backlight-smoother-steps", steps))
+		DBGLOG("igfx", "BLS: User requested steps = %u.", steps);
+	if (WIOKit::getOSDataValue(info->videoBuiltin, "backlight-smoother-interval", interval))
+		DBGLOG("igfx", "BLS: User requested interval = %u.", interval);
+	if (WIOKit::getOSDataValue(info->videoBuiltin, "backlight-smoother-threshold", threshold))
+		DBGLOG("igfx", "BLS: User requested threshold = %u.", threshold);
+	if (WIOKit::getOSDataValue(info->videoBuiltin, "backlight-smoother-queue-size", queueSize))
+		DBGLOG("igfx", "BLS: User requested queue size = %u.", queueSize);
+	if (WIOKit::getOSDataValue(info->videoBuiltin, "backlight-smoother-lowerbound", brightnessRange.first) ||
+		WIOKit::getOSDataValue(info->videoBuiltin, "backlight-smoother-upperbound", brightnessRange.second))
+		DBGLOG("igfx", "BLS: User requested brightness range = [%u, %u].", brightnessRange.first, brightnessRange.second);
+	
+	// Sanitize user configurations
+	if (steps == 0) {
+		SYSLOG("igfx", "BLS: Warning: User requested steps value is invalid. Will use the default value %u.", kDefaultSteps);
+		steps = kDefaultSteps;
+	}
+	
+	if (queueSize < kMinimumQueueSize) {
+		SYSLOG("igfx", "BLS: Warning: User requested queue size is too small. Will use the minimum value %u.", kMinimumQueueSize);
+		queueSize = kMinimumQueueSize;
+	}
+	
+	if (brightnessRange.first > brightnessRange.second) {
+		SYSLOG("igfx", "BLS: Warning: User requested brightness range is invalid. Will use the default range.");
+		brightnessRange.first = 0;
+		brightnessRange.second = UINT32_MAX;
+	}
+	
+	// Wrap this submodule as an OSObject
+	owner = OSObjectWrapper::with(this);
+	if (owner == nullptr) {
+		SYSLOG("igfx", "BLS: Failed to create the owner of the event source.");
+		goto error0;
+	}
+	
+	// Initialize the request queue
+	queue = BrightnessRequestQueue::withCapacity(queueSize);
+	if (queue == nullptr) {
+		SYSLOG("igfx", "BLS: Failed to initialize the request queue.");
+		goto error1;
```

Comment:
I strongly dislike these gotos, they make me feel there needs to be an outer function that will do failure printing and likely cleanup.

Reply:
I also hate them, but I do find them simplify error handling and clean-up code. Anyway, I revised the implementation of `deinit()` and replaced those `goto` statements with `deinit()` invocations.